### ISSUE TITLE
Thanks, one follow-on question, how to dynamically create the menu

### ISSUE THREAD

**Issue by sspilleman** [jwx/tutorials#3]
Hi!

Thanks for this great tutorial, it helped me a lot! I do have a question, playing around with this:

You are creating a fixed menu by manually adding the items to the nav. What I liked about the v1 router, is that I have a list in the router that I can iterate over to dynamically create the menu..
```html
<a repeat.for="row of router.navigation" if.to-view="row.settings.display" class="${row.isActive ? 'text-blue-800' : ''}" href="${row.href}">
  ${row.title}
</a>
```
How would I able to do it in v2. Is there a way to "inject the router" in the MyApp class or something alike and the iterate over the routes in the viewModel?

Thanks a lot!!

Sander

**Comment by sspilleman** [comment]
BTW, Im trying this and this doent't work. It does show the message, but it looks like a router object but its not the "current router"...

```javascript
import { Router } from "@aurelia/router";

export class MyApp {
    constructor(private router: Router) { }

    afterAttach() {
        console.log("after attached", this.router);
    }
}
```

**Comment by jwx** [comment]
Hello, @sspilleman!

If you're routing without configured routes there's no need to inject the router; any list will do to create dynamic menues. So something like
```typescript
const items = [
  { link: 'welcome', title: 'Welcome' },
  { link: 'about', title: 'About' },
];
```
```html
<a repeat.for="item of items" goto="${item.link}">${item.title}</a>
```
will do it. When you use the `goto` attribute, the class `goto-active` will be automatically added  to the active anchor element. What the list contains and how it's presented is up to you, so the `settings.display` example you had would work the same.

If you do want to inject the router, your example of that is correct except that you need to inject `IRouter`.

**Comment by sspilleman** [comment]
Hi Jürgen,

Thanks 4 your response. I did figure that I could create the array myself of course. I did like the fact that in v1 its already part of the router, and I didn't have to maintain it in a separate attribute myself.... hence my question.

If I change it to IRouter the object is empty "{}" in the afterAttach function. I tried both importing from "aurelia" and "@aurelia/router".... and routing IS working so it must have changed into something else again.
Also, in both cases I also tested to add a configureRouter method in my class... is isn't being called

Finally, I also tried adding a canActivate method to some other route/class/component, but that also isn't being called... 

You tutorial really got me started on getting basic routing to work, but other "(also)" basic stuff is hard to figure out :disappointed:

Sander

**Comment by jwx** [comment]
How do you get the menu into the router without maintaining/specifying it yourself in v1? If you want a more detailed specification you'll be able to get it through configured routes very similar to v1, but I don't understand how that's less to maintain than just a list of links and titles.

If you import and inject `IRouter` it should work. If it doesn't, please share code again.

The `configureRouter` is not needed or used in au2. (It might be added to support migration, but it's no longer needed since router configuration can be done everywhere by just injecting the router interface.

The `canActivate/activate` hooks are (currently, at least) renamed to `canEnter/enter`.

**Comment by sspilleman** [comment]
Because when the router is configured, you can get the routes from the router instead of putting them in a separate object. 

Yes it works now. I forgot to specifically used @inject, so it didn't DI-ed the instance, oops, my mistake. canEnter now also works!

I did see that in your example that he goto-active class on the books component view-model gets applied to all book-details  links, although only one book-details component is active if you select one (see screenshot, I have "book-details(1)" open in this case. I guess that is a bug?

Going further into this, Any pointers on how to implement:
1. Something similar to "mapUnknownRoutes" in v2? My current train of thought is to add a before hook, but then I need to maintain a list of all valid routes / states myself to check against, but that would be hard to maintain if the app grows.
2. Same for redirecting to a default route. Like in your demo, if you open the app, to redirect to "welcome" by default. I now have it implemented in a hook as follows. Is that the way to do it?
```javascript
if (navigatorInstruction.instruction === "" || navigatorInstruction.instruction === "/") {
    router.goto("welcome@main");
}
```

Tx, Sander

<img width="755" alt="Screenshot 2020-03-08 at 07 47 08" src="https://user-images.githubusercontent.com/592486/76157989-11d91880-6111-11ea-85c4-619f49261cbd.png">

**Comment by jwx** [comment]
@sspilleman Sorry for the delay.

If you're configuring routes, then sure, not having to specify the menu items again is an advantage. But if you're not configuring routes, I can see benefits of not having to involve the router in rendering a menu. Especially so if you have menus added a bit down in your component hierarchy.

Yeah, the multiple `goto-active` was a bug. Somehow a change I made locally hadn't made it into master. I made a PR that was merged earlier tonight though so it should be working now.

As for your implementation questions, I've made a new video: https://youtu.be/F8yjcIe4Hhg

Let me know if that answers your questions.
